Tribes/src/app/_services/building.service.ts

```diff
@@ -27,14 +29,52 @@ export class BuildingService {
   }
 
   addNewBuilding(type: string): void {
+    this.updateRessourceByConstruction.emit(COST_NEW_BUILDING);
     this.http.post<Building>(ROOT_URL + '/kingdom/buildings', {type})
     .pipe(catchError(this.errorHandlingService.handleError))
     .subscribe(response => this.handleBuildingProcess(response),
               error => console.error(error));
   }
 
-  handleBuildingProcess(building: Building) {
+  handleBuildingProcess(building: Building): void {
     this.beginConstruction.emit();
-    setTimeout(() => {this.finishConstruction.emit(); }, building.finishedAt - building.startedAt);
+    this.saveFinishedAtInLocalStorage(building.finishedAt);
+    setTimeout(() => { console.warn(building.id);
+      this.finishConstruction.emit(); }, building.finishedAt - building.startedAt);
+  }
+
+  upgradeBuilding(idToUpgrade: number, level: number): void {
+    this.updateRessourceByConstruction.emit(COST_BASE_UPGRADE * (level - 1));
+    this.http.put<Building>(ROOT_URL + '/kingdom/buildings/' + idToUpgrade, { level })
+    .pipe(catchError(this.errorHandlingService.handleError))
+    .subscribe(response => this.handleBuildingProcess(response),
+              error => console.error(error));
+  }
+
+  checkIfBuildingIsProgressing(building: Building): boolean {
+    const currentTime = new Date().getTime();
+    return currentTime <= building.finishedAt;
+  }
+
+  saveFinishedAtInLocalStorage(finishedAt: number) {
+    if (localStorage.getItem(this.keyTimeouts) === null) {
+      localStorage.setItem(this.keyTimeouts, JSON.stringify([finishedAt]));
+    } else {
+      let savedTimeouts = [];
+      const now = new Date().getTime();
+      savedTimeouts = JSON.parse(localStorage.getItem(this.keyTimeouts));
+      savedTimeouts = savedTimeouts.filter(timeout => (timeout > now));
+      savedTimeouts.push(finishedAt);
+      localStorage.setItem(this.keyTimeouts, JSON.stringify(savedTimeouts));
+    }
+  }
+
+  setTimeoutsAgain() {
+    if (localStorage.getItem(this.keyTimeouts)) {
+      let savedTimeouts = [];
+      savedTimeouts = JSON.parse(localStorage.getItem(this.keyTimeouts)).filter(timeout => (timeout > new Date().getTime()));
+      savedTimeouts.forEach(timeout => { setTimeout(() => { this.finishConstruction.emit(); }, timeout - new Date().getTime()); });
```

Comment:
This looks good enough for now, if it works.

However I'd move the whole out to an external function like `setFinishConstructionTimeout` that gets the time as a parameter. By the way I just noticed, that you do not save which building had the timeout. I assume when this event happens, you just remove one randomly? Or what minds this? Why do you have this in the building service, and why not in the progress bar is checking this?

Reply:
It's not randomly, the function will remove the building, which is finished, it's enough to know the finishedAt property of the building, the id is not necessary. 
There is more components, that are subsrcibe to the finishConstruction event, and this function is called, when we buy a new building or upgrade one, and this functions belong to the building.service.